throw remote exception on client side

zhaojuanmao opened the pull request:
Summary: catch exception thrown on server, send the exception message back to client and rethrow it.

1. wrapped exception + traceback thrown in python, passed them back as python result string and then threw them on client.
2. C++ exception are passed as exception message type and then threw them on client

Differential Revision: D16748748

xush6528 commented on test/test_rpc.py:
```diff
@@ -198,12 +198,11 @@ def test_py_no_return_result(self):
-            expected = "run_python_udf_internal caught exception: " + str(e)
-        self.assertEqual(ret, expected)
+            ret = str(e)
+        self.assertTrue(isinstance(ret, str))
```
Use `self.assertRaisesRegex`?

https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaisesRegex

xush6528 commented on torch/distributed/internal_rpc_utils.py:
```diff
@@ -13,11 +13,8 @@ def serialize(obj):
-    try:
-        python_udf = pickle.loads(pickled_python_udf)
-        result = python_udf.func(*python_udf.args, **python_udf.kwargs)
-    except Exception as e:
```
Awesome!

xush6528 reviewed (commented):
Thanks for this work. This will help developers a lot!

aazzolini reviewed (commented):
Could you paste a print of how the exception gets propagated to the client? Does it show the stack trace of where the exception happened on the server? It's very important to have fully traceable exceptions for proper debugging.

zhaojuanmao commented:
This is the exception message thrown on client for the unit test. I think it captures the errors thrown by python UDF call itself, because I switched using CPython API to pybind11 API that can give much better exception handling. 

please let me know whether the message is what you want

At:
  /data/users/yanlizhao/fbsource/fbcode/buck-out/dev/gen/caffe2/test/rpc#binary,link-tree/torch/distributed/internal_rpc_utils.py(17): run_python_udf_internal
TypeError: no_result() takes 0 positional arguments but 1 was given

xush6528 commented:
I am afraid this error message is not sufficient for developers to debug.

Imainge you have multiple calls that might throw exception.

This exception doesn't tell developers which line is "the call" that triggers the exception. 

----------------------
Also, it would be better if we have add another test case for errors that raise in user function.

```Python
def raise_func(self):
    raise ValueError("Expected error")

def test_raise_in_user_func(self):
    fut = dist.rpc(async_call)
    with self. assertRaisesRegex(ValueError, "Expected error.+Traceback:"):
         fut.get()
```

zhaojuanmao commented:
@xush6528 right, it does not tell the line, but it tells which call failed. e.g. it said no_result() call failed because of what.

I think it just propagates whatever python call throws

aazzolini commented:
@zhaojuanmao , I believe it is not printing the traceback -- that's fine for a simple handler like "no_return" but it may be hard to debug exceptions thrown more internally when an RPC handler calls into more complex code. It would be interesting if you could extract the traceback from python and serialize it over the wire as well. This will save us a lot of debugging time.

zhaojuanmao commented:
@aazzolini and @xush6528, thanks for the suggestion.
 1. wrapped exception + traceback thrown in python, passed them back as python result string and then threw them on client. 
 2. C++ exception are passed as exception message type and then threw them on client
 3. added one more test case. 
 4. also printed out the exception thrown on client side, please let me know whether the info is sufficient:

**test_py_function_exception**

ValueError('Expected error',)
Traceback (most recent call last):
  File "/data/users/yanlizhao/fbsource/fbcode/buck-out/dev/gen/caffe2/test/rpc#binary,link-tree/torch/distributed/internal_rpc_utils.py", line 20, in run_python_udf_internal
    result = python_udf.func(*python_udf.args, **python_udf.kwargs)
  File "/data/users/yanlizhao/fbsource/fbcode/buck-out/dev/gen/caffe2/test/rpc#binary,link-tree/test_rpc.py", line 22, in raise_func
    raise ValueError("Expected error")
ValueError: Expected error


**test_py_raise_in_user_func**

TypeError('no_result() takes 0 positional arguments but 1 was given',)
Traceback (most recent call last):
  File "/data/users/yanlizhao/fbsource/fbcode/buck-out/dev/gen/caffe2/test/rpc#binary,link-tree/torch/distributed/internal_rpc_utils.py", line 20, in run_python_udf_internal
    result = python_udf.func(*python_udf.args, **python_udf.kwargs)
TypeError: no_result() takes 0 positional arguments but 1 was given

aazzolini commented:
Could you provide another example where the exception is thrown in user code? Notice that in the example that you provided above, the exception is thrown in pytorch code.

zhaojuanmao commented:
@aazzolini if I try catch the exception in test/user codes and print(str(e)),  the exception message is the same as what I pasted before, see as follows. What kind of different info do you expect?

**test_py_raise_in_user_func**
ValueError('Expected error',)
Traceback (most recent call last):
  File "/data/users/yanlizhao/fbsource/fbcode/buck-out/dev/gen/caffe2/test/rpc#binary,link-tree/torch/distributed/internal_rpc_utils.py", line 19, in run_python_udf_internal
    result = python_udf.func(*python_udf.args, **python_udf.kwargs)
  File "/data/users/yanlizhao/fbsource/fbcode/buck-out/dev/gen/caffe2/test/rpc#binary,link-tree/test_rpc.py", line 24, in raise_func
    raise ValueError("Expected error")
ValueError: Expected error

**test_py_function_exception**
TypeError('no_result() takes 0 positional arguments but 1 was given',)
Traceback (most recent call last):
  File "/data/users/yanlizhao/fbsource/fbcode/buck-out/dev/gen/caffe2/test/rpc#binary,link-tree/torch/distributed/internal_rpc_utils.py", line 19, in run_python_udf_internal
    result = python_udf.func(*python_udf.args, **python_udf.kwargs)
TypeError: no_result() takes 0 positional arguments but 1 was given

mrshenli reviewed (approved):
LGTM!

xush6528 reviewed (approved):
I saw 
> test_rpc.py", line 22, in raise_func

I think it's good now.

xush6528 commented on test/test_rpc.py:
```diff
@@ -214,13 +217,17 @@ def test_py_no_return_result(self):
     @_wrap_with_rpc
     def test_py_function_exception(self):
```
I wonder is it possible to get the Python line number about this error too.

You happened to find a good test case. Users passing wrong argument is a common error.

```
def test_py_user_pass_wrong_arguments_exception(self):
```

xush6528 commented:
I see, from the current error trace, the printed lines are only about code that run by the callee side.
The callee does not know which line the caller was at. 

In order to know the caller side stack trace,
I think one solution is, on invoking an RPC, the caller side saves the stack trace of itself together with the request future, if the response is an exception, the caller side pulls out the stack trace on itself and concatenate with the other stack trace on the callee side.

This could be a big change, we could do it in a later PR.


For example, put the call site on itself together with the future.

pritamdamania87 commented on torch/csrc/distributed/rpc/functions.cpp:
```diff
@@ -8,28 +8,56 @@ void processRequestBlocking(
+            from,
+            Message(
+                std::move(payload),
+                std::move(tensorTable),
```
nit: call use `std::vector<torch::Tensor>()` here instead of a separate `tensorTable` var.

pritamdamania87 commented on torch/csrc/distributed/rpc/functions.cpp:
```diff
@@ -8,28 +8,56 @@ void processRequestBlocking(
+        agent.send(from, std::move(response));
+      } catch (std::exception& e) {
+        std::vector<torch::Tensor> tensorTable;
+        std::vector<char> payload;
```
nit: `std::vector<char> payload(err, err + strlen(err))`.

pritamdamania87 commented on torch/csrc/distributed/rpc/functions.cpp:
```diff
@@ -8,28 +8,56 @@ void processRequestBlocking(
+                std::move(tensorTable),
+                MessageType::PYTHON_RET,
+                request.id()));
+      } catch (std::exception& e) {
```
Can we de-dup some of this error checking code? Its pretty similar to what we have in `case MessageType::SCRIPT_CALL` above.

pritamdamania87 reviewed (changes requested):
Requesting changes mostly to address the duplicate error checking code.

zhaojuanmao committed:
throw remote exception on client side (#24138)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/24138

catch exception thrown on server, send the exception message back to client and rethrow it.

Reviewed By: mrshenli

Differential Revision: D16748748

fbshipit-source-id: af9a871e0e4c183de5105c993a8444588658657d

zhaojuanmao commented:
addressed @pritamdamania's comments, test failures are not related. 
@xush6528 got your point, agree that client/caller side traceback especially C++ traceback possibly is tricky to catch 

landing
